spring-cloud-zookeeper-discovery/src/main/java/org/springframework/cloud/zookeeper/discovery/watcher/DependencyStateChangeListenerRegistry.java

```diff
@@ -0,0 +1,47 @@
+package org.springframework.cloud.zookeeper.discovery.watcher;
+
+import org.apache.curator.framework.CuratorFramework;
+import org.apache.curator.framework.state.ConnectionState;
+import org.apache.curator.x.discovery.ServiceCache;
+import org.apache.curator.x.discovery.details.ServiceCacheListener;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.List;
+
+public class DependencyStateChangeListenerRegistry implements ServiceCacheListener {
+
+    private static final Logger log = LoggerFactory.getLogger(DependencyStateChangeListenerRegistry.class);
+
```

Comment:
This class isn't thread safe. The `listeners` array list is modified upon startup by the thread starting the application context and iterated by the Curator listener thread. It may also be modified via the public `DependencyWatcher` methods it implements.